fedimint-core/src/db/mod.rs

```diff
@@ -437,43 +541,47 @@ impl<'isolated, 'parent, T: Send + Encodable + 'isolated> IDatabaseTransaction<'
             .await
     }
 
-    async fn raw_find_by_prefix(&mut self, key_prefix: &[u8]) -> PrefixStream<'_> {
+    async fn raw_find_by_prefix(&mut self, key_prefix: &[u8]) -> Result<PrefixStream<'_>> {
         let mut prefix_with_module = self.prefix.clone();
         prefix_with_module.extend_from_slice(key_prefix);
         let raw_prefix = self
             .inner_tx
             .raw_find_by_prefix(prefix_with_module.as_slice())
-            .await;
+            .await?;
 
-        Box::pin(raw_prefix.map(|kv| {
+        Ok(Box::pin(raw_prefix.map(|kv| {
             let key = kv.0;
             let stripped_key = &key[(self.prefix.len())..];
             (stripped_key.to_vec(), kv.1)
-        }))
+        })))
+    }
+
+    async fn raw_remove_by_prefix(&mut self, key_prefix: &[u8]) -> Result<()> {
+        self.inner_tx.raw_remove_by_prefix(key_prefix).await
     }
 
-    async fn commit_tx(self: Box<Self>) -> Result<()> {
+    async fn commit_tx(&mut self) -> Result<()> {
         panic!("DatabaseTransaction inside modules cannot be committed");
     }
 
-    async fn rollback_tx_to_savepoint(&mut self) {
+    async fn rollback_tx_to_savepoint(&mut self) -> Result<()> {
         self.inner_tx.rollback_tx_to_savepoint().await
     }
 
-    async fn set_tx_savepoint(&mut self) {
+    async fn set_tx_savepoint(&mut self) -> Result<()> {
         self.inner_tx.set_tx_savepoint().await
     }
 }
 
 #[doc = " A handle to a type-erased database implementation"]
 pub struct DatabaseTransaction<'a> {
-    tx: Box<dyn IDatabaseTransaction<'a>>,
+    tx: Box<dyn ISingleUseDatabaseTransaction<'a>>,
     decoders: ModuleDecoderRegistry,
     commit_tracker: CommitTracker,
 }
 
 impl<'a> std::ops::Deref for DatabaseTransaction<'a> {
-    type Target = dyn IDatabaseTransaction<'a>;
+    type Target = dyn ISingleUseDatabaseTransaction<'a>;
```

Comment:
this is not a good idea. This adds `DatabaseTransaction::commit` which takes `&mut self` and leaves DatabaseTransaction in unusable state. 
`DatabaseTransaction` should implement its own `commit(self)`

Reply:
Removed the deref altogether